src/main/java/org/truffleruby/language/loader/FeatureLoader.java

```diff
@@ -50,10 +56,35 @@
     private TruffleObject sulongLoadLibraryFunction;
     private Map<String, String> nativeLibraryMap;
 
+    private NativeFunction getcwd;
+    private static final int PATH_MAX = 1024; // jnr-posix hard codes this value
+    private int cwdBufferSize = PATH_MAX;
+    private Pointer cwdBuffer;
+
     public FeatureLoader(RubyContext context) {
         this.context = context;
     }
 
+    public void initialize(NativePlatform nativePlatform, FinalizationService finalizationService) {
+        final TruffleNFIPlatform nfi = nativePlatform.getTruffleNFI();
+        final Object size_t_typedef = nativePlatform.getRubiniusConfiguration().get("rbx.platform.typedef.size_t");
+        final String size_t = nfi.toNFIType(StringOperations.getString((DynamicObject) size_t_typedef));
+
+        this.getcwd = nfi.getFunction("getcwd", 2, "(pointer," + size_t + "):pointer");
+        this.cwdBuffer = Pointer.malloc(cwdBufferSize);
+        this.cwdBuffer.enableAutorelease(finalizationService);
+    }
+
+    private String getWorkingDirectory() {
+        final TruffleNFIPlatform nfi = context.getNativePlatform().getTruffleNFI();
+        final long address = nfi.asPointer((TruffleObject) getcwd.call(cwdBuffer.address(), cwdBufferSize));
+        if (address == 0) {
+            throw new UnsupportedOperationException("getcwd() failed");
+        }
+        final byte[] bytes = cwdBuffer.readZeroTerminatedByteArray(0);
+        return new String(bytes, StandardCharsets.UTF_8);
```

Comment:
I'm not sure if this should be the locale or the external encoding, but I don't think it should be hard-coded to UTF-8.

Reply:
Every jcodings `Encoding` object has a `getCharset` method, but we can't PE that. You might need to add something to `StringOperations`. We have code close to what you'd need in `RopeOperations`, but I'm not seeing anything with the exact signature necessary.

Reply:
Right, I forgot to fix this, will do.

Reply:
I think the JNR call used UTF-8 no matter what the locale yes, let's see if using the locale encoding works.

Reply:
It gets tricky when creating Java strings, for sure. Internally, the String constructor does some decoding.